File: projects/basis/sdk-review/py-v2/src/modules/Taxes.ts
import { BasisClient } from '../BasisClient';
import ATaxesArtifact from '../abis/ATaxes.json';
import { Address } from 'viem';

export class TaxesModule {
  private client: BasisClient;
  private taxesAddress: Address;

  constructor(client: BasisClient, taxesAddress: Address) {
    this.client = client;
    this.taxesAddress = taxesAddress;
  }

  private async _syncTx(txHash: string) {
    try {
      await this.client.api.syncTransaction(txHash);
    } catch (e: any) {
      console.warn('Sync warning:', e.message || e);
    }
  }

  /**
   * Returns the effective tax rate (in basis points) for a specific token and user.
   */
  async getTaxRate(token: Address, user: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.taxesAddress,
      abi: ATaxesArtifact.abi,
      functionName: 'getTaxRate',
      args: [token, user],
    }) as Promise<bigint>;
  }

  /**
   * Returns the current surge tax rate (in basis points) for a token.
   */
  async getCurrentSurgeTax(token: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.taxesAddress,
      abi: ATaxesArtifact.abi,
      functionName: 'getCurrentSurgeTax',
      args: [token],
    }) as Promise<bigint>;
  }

  /**
   * Returns the available surge quota for a token.
   */
  async getAvailableSurgeQuota(token: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.taxesAddress,
      abi: ATaxesArtifact.abi,
      functionName: 'availableSurgeQuota',
      args: [token],
    }) as Promise<bigint>;
  }

  /**
   * Returns all four base tax rates.
   */
  async getBaseTaxRates() {
    const [stasis, stable, defaultRate, prediction] = await Promise.all([
      this.client.publicClient.readContract({
        address: this.taxesAddress,
        abi: ATaxesArtifact.abi,
        functionName: '_taxRateStasis',
      }) as Promise<bigint>,
      this.client.publicClient.readContract({
        address: this.taxesAddress,
        abi: ATaxesArtifact.abi,
        functionName: '_taxRateStable',
      }) as Promise<bigint>,
      this.client.publicClient.readContract({
        address: this.taxesAddress,
        abi: ATaxesArtifact.abi,
        functionName: '_taxRateDefault',
      }) as Promise<bigint>,
      this.client.publicClient.readContract({
        address: this.taxesAddress,
        abi: ATaxesArtifact.abi,
        functionName: '_taxRatePrediction',
      }) as Promise<bigint>,
    ]);

    return { stasis, stable, default: defaultRate, prediction };
  }

  /**
   * Start a decaying surge tax on a factory token. Only callable by the token's DEV.
   */
  async startSurgeTax(startRate: bigint, endRate: bigint, duration: bigint, token: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required.");
    }
    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.taxesAddress,
      abi: ATaxesArtifact.abi,
      functionName: 'startSurgeTax',
      args: [startRate, endRate, duration, token],
    });
    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * End an active surge tax early. Only callable by the token's DEV.
   */
  async endSurgeTax(token: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required.");
    }
    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.taxesAddress,
      abi: ATaxesArtifact.abi,
      functionName: 'endSurgeTax',
      args: [token],
    });
    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Add a developer revenue share wallet for a token. Only callable by the token's DEV.
   */
  async addDevShare(token: Address, wallet: Address, basisPoints: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required.");
    }
    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.taxesAddress,
      abi: ATaxesArtifact.abi,
      functionName: 'addDevShare',
      args: [token, wallet, basisPoints],
    });
    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Remove a developer revenue share wallet. Only callable by the token's DEV.
   */
  async removeDevShare(token: Address, wallet: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required.");
    }
    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.taxesAddress,
      abi: ATaxesArtifact.abi,
      functionName: 'removeDevShare',
      args: [token, wallet],
    });
    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return { hash, receipt };
  }
}

File: projects/basis/sdk-review/python/basis/modules/taxes.py
from web3 import Web3
from .factory import load_abi


class TaxesModule:
    def __init__(self, client, taxes_address: str):
        self.client = client
        self.taxes_address = Web3.to_checksum_address(taxes_address)
        self.taxes_abi = load_abi('ATaxes.json')
        self.contract = self.client.web3.eth.contract(address=self.taxes_address, abi=self.taxes_abi)

    # ------------------------------------------------------------------
    # Read methods
    # ------------------------------------------------------------------

    def get_tax_rate(self, token: str, user: str) -> int:
        """Returns the tax rate in basis points for a token/user pair."""
        checksum_token = Web3.to_checksum_address(token)
        checksum_user = Web3.to_checksum_address(user)
        return self.contract.functions.getTaxRate(checksum_token, checksum_user).call()

    def get_current_surge_tax(self, token: str) -> int:
        """Returns the current surge tax for a token."""
        checksum_token = Web3.to_checksum_address(token)
        return self.contract.functions.getCurrentSurgeTax(checksum_token).call()

    def get_available_surge_quota(self, token: str) -> int:
        """Returns the available surge quota for a token."""
        checksum_token = Web3.to_checksum_address(token)
        return self.contract.functions.availableSurgeQuota(checksum_token).call()

    def get_base_tax_rates(self) -> dict:
        """Returns the base tax rates for all token types."""
        stasis = self.contract.functions._taxRateStasis().call()
        stable = self.contract.functions._taxRateStable().call()
        default = self.contract.functions._taxRateDefault().call()
        prediction = self.contract.functions._taxRatePrediction().call()
        return {
            'stasis': stasis,
            'stable': stable,
            'default': default,
            'prediction': prediction,
        }

    # ------------------------------------------------------------------
    # Write methods
    # ------------------------------------------------------------------

    def _build_and_send_tx(self, function_call):
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")
        tx = function_call.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
        })
        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)
        return {'hash': tx_hash.hex(), 'receipt': receipt}

    def start_surge_tax(self, start_rate: int, end_rate: int, duration: int, token: str):
        """Start a decaying surge tax on a factory token. Only callable by the token's DEV."""
        func = self.contract.functions.startSurgeTax(start_rate, end_rate, duration, Web3.to_checksum_address(token))
        return self._build_and_send_tx(func)

    def end_surge_tax(self, token: str):
        """End an active surge tax early. Only callable by the token's DEV."""
        func = self.contract.functions.endSurgeTax(Web3.to_checksum_address(token))
        return self._build_and_send_tx(func)

    def add_dev_share(self, token: str, wallet: str, basis_points: int):
        """Add a developer revenue share wallet. Only callable by the token's DEV."""
        func = self.contract.functions.addDevShare(
            Web3.to_checksum_address(token), Web3.to_checksum_address(wallet), basis_points
        )
        return self._build_and_send_tx(func)

    def remove_dev_share(self, token: str, wallet: str):
        """Remove a developer revenue share wallet. Only callable by the token's DEV."""
        func = self.contract.functions.removeDevShare(
            Web3.to_checksum_address(token), Web3.to_checksum_address(wallet)
        )
        return self._build_and_send_tx(func)
